evm-contracts/src/v0.7/dev/StalenessFlaggingValidator.sol

```diff
@@ -0,0 +1,164 @@
+  );
+
+  /**
+   * @notice Create a new StalenessFlaggingValidator
+   * @param flagsAddress Address of the flag contract
+   * @dev Ensure that this contract has sufficient write permissions
+   * on the flag contract
+   */
+  constructor(address flagsAddress) 
+    public 
+    ConfirmedOwner(msg.sender)
+  {
+    setFlagsAddress(flagsAddress);
+  }
+
+  /**
+   * @notice Updates the flagging contract address for raising flags
+   * @param flagsAddress sets the address of the flags contract
+   */
+  function setFlagsAddress(address flagsAddress)
+    public
+    onlyOwner()
+  {
+    address previous = address(s_flags);
+    if (previous != flagsAddress) {
+      s_flags = FlagsInterface(flagsAddress);
+      emit FlagsAddressUpdated(previous, flagsAddress);
+    }
+  }
+
+  /**
+   * @notice Set the threshold limits for each aggregator
+   * @dev parameters must be same length
+   * @param aggregators address[] memory
+   * @param flaggingThresholds uint256[] memory
+   */
+  function setThresholds(address[] memory aggregators, uint256[] memory flaggingThresholds)
+    public 
+    onlyOwner()
+  {
+    require(aggregators.length == flaggingThresholds.length, "Different sized arrays");
+    for (uint256 i = 0; i < aggregators.length; i++) {
+      address aggregator = aggregators[i];
+      uint256 previousThreshold = s_thresholdSeconds[aggregator];
+      uint newThreshold = flaggingThresholds[i];
+      if (previousThreshold != newThreshold) {
+        s_thresholdSeconds[aggregator] = newThreshold;
+        emit FlaggingThresholdUpdated(aggregator, previousThreshold, newThreshold);
+      }
+    }
+  }
+
+  /**
+   * @notice Check for staleness in an array of aggregators
+   * @dev If any of the aggregators are stale, this function will return true,
+   * otherwise false
+   * @param aggregators address[] memory
+   * @return bool
+   */
+  function check(address[] memory aggregators) public view returns (bool) {
+    uint256 currentTimestamp = block.timestamp;
+
+    for (uint256 i = 0; i < aggregators.length; i++) {
+      if (isStale(aggregators[i], currentTimestamp)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * @notice Check for staleness in an array of aggregators, raise a flag
+   * on the flags contract for each aggregator that is stale
+   * @dev This contract must have write permissions on the flags contract
+   * @param aggregators address[] calldata
+   */
+  function update(address[] calldata aggregators) external returns (address[] memory){
+    uint256 currentTimestamp = block.timestamp;
+
+    uint256 aggregatorsLength = aggregators.length;
+    address[] memory staleAggregators = new address[](aggregatorsLength);
+    uint256 staleCount = 0;
+    for (uint256 i = 0; i < aggregatorsLength; i++) {
+      address aggregator = aggregators[i];
+      if (isStale(aggregator, currentTimestamp)) {
+        staleAggregators[staleCount] = aggregator;
+        staleCount++;
+      }
+    }
+
```

Comment:
Using assembly to reduce the length of the array passed into `raiseFlags`.

@se3000